sbol2/document.py

```diff
@@ -581,13 +586,72 @@ def parse_properties_inner(self, subject, predicate, obj):
                 msg = msg.format(subject, type(subject))
                 self.logger.debug(msg)
 
-    def remove_descendants(self):
-        to_delete = []
-        for name, obj in self.SBOLObjects.items():
-            if not obj.is_top_level():
-                to_delete.append(name)
-        for name in to_delete:
-            del self.SBOLObjects[name]
+    def find_reference(self, uri):
+        """Find objects that reference the given URI. Returns a list of
+        objects. The list will be empty if no references were found.
+
+        """
+        references = []
+        for obj in self.SBOLObjects.values():
+            references.extend(obj.find_reference(uri))
+        return references
+
+    def parse_annotation_objects(self):
+        """Parse leftover objects from reading and link them up where they
+        belong. These are usually extension-type objects.
+
+        """
+        for uri, obj in self.SBOLObjects.items():
+            self.logger.debug('Possible annotation object %s', obj.identity)
+        annotation_objects = [obj for obj in self.SBOLObjects.values()
+                              if not isinstance(obj, TopLevel)]
+        for ao in annotation_objects:
+            self.logger.debug('Annotation object: %s', ao.identity)
+            if SBOL_PERSISTENT_IDENTITY in ao.properties:
+                # Copy to a new TopLevel object
+                tl = TopLevel(type_uri=ao.type)
+                for k, v in ao.properties.items():
+                    tl.properties[k] = v
+                for k, v in ao.owned_objects.items():
+                    tl.owned_objects[k] = v
+                tl.doc = self
```

Comment:
This is a case where the `cast` method would be helpful, e.g.:
```
tl = ao.cast(TopLevel)
```
No matter, this code block accomplishes the same thing.

Reply:
One of us is confused. I don't think the `cast` operator creates a new object. The C++ code [creates a new object in this case](https://github.com/SynBioDex/libSBOL/blob/master/source/document.cpp#L895-L903), and this python code is a direct translation of that. The C++ code does not use the cast operator here.

Reply:
The point I'm trying to make is that this pattern tends to show up in different places.  This code creates a new object of type TopLevel, then copies all the properties from the base SBOLObject, and then replaces that SBOLObject (bye, bye!) with the new TopLevel.  Effectively, that means you have downcast from an object of type SBOLObject to TopLevel.  That's what the `cast` methods were implemented for:
https://github.com/SynBioDex/libSBOL/blob/develop/source/object.cpp#L78
Just saying, it's a pattern that shows up now and again, where the `cast` method can be used to accomplish the same thing in a more explicit manner.  This is more of an FYI than a request for a code change.

Reply:
❗  I did not know that. I thought cast was doing something with the underlying libSBOL object. Namely, just casting it to a different type. I didn't know it was actually creating a new object. It's more like a copy than a cast. Maybe.